urlcdd0c1ab49_emiln_JHouseFireExplorer_tgz-pJ8-dk_brics_automaton_DatatypesJ8/src/dk/brics/automaton/ShuffleOperations.java

```diff
@@ -0,0 +1,338 @@
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Set;
+
+/**
+ * Automata operations involving shuffling.
+ */
+final public class ShuffleOperations {
+	
+	private ShuffleOperations() {}
+
+	/** 
+	 * Returns an automaton that accepts the shuffle (interleaving) of 
+	 * the languages of the given automata.
+	 * As a side-effect, both automata are determinized, if not already deterministic.     
+	 * Never modifies the input automata languages.
+	 * <p>
+	 * Complexity: quadratic in number of states (if already deterministic). 
+	 * <p>
+	 * <dl><dt><b>Author:</b></dt><dd>Torben Ruby 
+	 * &lt;<a href="mailto:ruby@daimi.au.dk">ruby@daimi.au.dk</a>&gt;</dd></dl>
+	 */
+	public static Automaton shuffle(Automaton a1, Automaton a2) {
+		a1.determinize();
+		a2.determinize();
+		Transition[][] transitions1 = Automaton.getSortedTransitions(a1.getStates());
+		Transition[][] transitions2 = Automaton.getSortedTransitions(a2.getStates());
+		Automaton c = new Automaton();
+		LinkedList<StatePair> worklist = new LinkedList<StatePair>();
+		HashMap<StatePair, StatePair> newstates = new HashMap<StatePair, StatePair>();
+		State s = new State();
+		c.initial = s;
+		StatePair p = new StatePair(s, a1.initial, a2.initial);
+		worklist.add(p);
+		newstates.put(p, p);
+		while (worklist.size() > 0) {
+			p = worklist.removeFirst();
+			p.s.accept = p.s1.accept && p.s2.accept;
+			Transition[] t1 = transitions1[p.s1.number];
+			for (int n1 = 0; n1 < t1.length; n1++) {
+				StatePair q = new StatePair(t1[n1].to, p.s2);
+				StatePair r = newstates.get(q);
+				if (r == null) {
+					q.s = new State();
+					worklist.add(q);
+					newstates.put(q, q);
+					r = q;
+				}
+				p.s.transitions.add(new Transition(t1[n1].min, t1[n1].max, r.s));
+			}
+			Transition[] t2 = transitions2[p.s2.number];
+			for (int n2 = 0; n2 < t2.length; n2++) {
+				StatePair q = new StatePair(p.s1, t2[n2].to);
+				StatePair r = newstates.get(q);
+				if (r == null) {
+					q.s = new State();
+					worklist.add(q);
+					newstates.put(q, q);
+					r = q;
+				}
+				p.s.transitions.add(new Transition(t2[n2].min, t2[n2].max, r.s));
+			}
+		}
+		c.deterministic = false;
+		c.removeDeadTransitions();
+		c.checkMinimizeAlways();
+		return c;
+	}
+	
+	/**
+	 * Returns a string that is an interleaving of strings that are accepted by
+	 * <code>ca</code> but not by <code>a</code>. If no such string
+	 * exists, null is returned. As a side-effect, <code>a</code> is determinized, 
+	 * if not already deterministic. Only interleavings that respect
+	 * the suspend/resume markers (two BMP private code points) are considered if the markers are non-null. 
+	 * Also, interleavings never split surrogate pairs.
+	 * <p>
+	 * Complexity: proportional to the product of the numbers of states (if <code>a</code>
+	 * is already deterministic).
+	 */ 
+	public static String shuffleSubsetOf(Collection<Automaton> ca, Automaton a, Character suspend_shuffle, Character resume_shuffle) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 34. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 120-130 into a separate method.